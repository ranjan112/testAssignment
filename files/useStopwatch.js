import { useState, useEffect, useRef } from "react";

const useStopwatch = () => {
    const [runningTime, setRunningTime] = useState(0);
  const [startTime, setStartTime] = useState(0);
  const [isRunning, setIsRunning] = useState(false);
  const [laps, setLaps] = useState([]);
  const intervalRef = useRef(null);
  const pauseTimeRef = useRef(0);

  const startStopwatch = () => {
    if (runningTime === 0) {
      setStartTime(performance.now());
    } else {
      const pauseDuration = performance.now() - pauseTimeRef.current;
      setStartTime((prevStartTime) => prevStartTime + pauseDuration);
    }
    setIsRunning(true);
  };

  const pauseStopwatch = () => {
    setIsRunning(false);
    pauseTimeRef.current = performance.now();
  };

  const resetStopwatch = () => {
    setRunningTime(0);
    setIsRunning(false);
    setLaps([]);
  };

  const lapStopwatch = () => {
    setLaps((prevLaps) => {
      return [runningTime, ...prevLaps];
    });
  };

  useEffect(() => {
    if (isRunning) {
      intervalRef.current = setInterval(() => {
        setRunningTime(performance.now() - startTime);
      }, 10);
    } else {
      clearInterval(intervalRef.current);
    }

    return () => clearInterval(intervalRef.current);
  }, [isRunning, startTime]);


  const formatTime = (time) => {
    const totalSeconds = Math.floor(time / 1000);
    const hours = Math.floor(totalSeconds / 3600);
    const minutes = Math.floor((totalSeconds % 3600) / 60);
    const seconds = totalSeconds % 60;
    const milliseconds = Math.floor(time % 1000 / 10);
    const formattedHours = hours.toString().padStart(2, "0");
    const formattedMinutes = minutes.toString().padStart(2, "0");
    const formattedSeconds = seconds.toString().padStart(2, "0");
    const formattedMilliseconds = milliseconds.toString().padStart(2, "0");

    if (hours > 0) {
      return `${formattedHours}:${formattedMinutes}:${formattedSeconds}.${formattedMilliseconds}`;
    }

    return `${formattedMinutes}:${formattedSeconds}.${formattedMilliseconds}`;
  };

  return {
    timeRemaining: formatTime(runningTime),
    isRunning,
    laps,
    startStopwatch,
    pauseStopwatch,
    resetStopwatch,
    lapStopwatch,
    formatTime,
  };
};

export default useStopwatch;