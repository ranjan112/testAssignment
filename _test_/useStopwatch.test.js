import { renderHook, act } from '@testing-library/react-hooks';
import useStopwatch from '../files/useStopwatch';
import renderer from "react-test-renderer";

jest.useFakeTimers();

describe('useTimer', () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  //snapshot testing done

  describe("Jest Snapshot testing suite", () => {
    it("Matches DOM Snapshot", () => {
      const domTree = renderer.create(<useStopwatch />).toJSON();
      expect(domTree).toMatchSnapshot();
    });
  });

  //checking function test on start of stopwatch time
  it('it will start and pause the timer watch', () => {
    const startStopwatch = jest.fn();
    const { result } = renderHook(() => useStopwatch(10, startStopwatch));

    act(() => {
      result.current.startStopwatch();
    });

    expect(result.current.isRunning).toBe(true);
    act(() => {
        result.current.pauseStopwatch();
      });
  
      expect(result.current.isRunning).toBe(false);

});
//checking function test on reset of stopwatch time
it('it will reset the timer correctly', () => {
    const onTimerEnd = jest.fn();
    const { result } = renderHook(() => useStopwatch(10, onTimerEnd));

    act(() => {
      result.current.startStopwatch();
    });

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    act(() => {
      result.current.resetStopwatch();
    });

    expect(result.current.isRunning).toBe(false);
  });

  //checking function test on end of stopwatch time
  it('should call onTimerEnd when the timer ends', () => {
    const onTimerEnd = jest.fn();
    const { result } = renderHook(() => useStopwatch(1, onTimerEnd));

    act(() => {
      result.current.startStopwatch();
    });

    act(() => {
      jest.advanceTimersByTime(1000);
    });

  });

  //checking function test on lapse of stopwatch time
  it('should call laps when the timer ', () => {
    const onTimerEnd = jest.fn();
    const { result } = renderHook(() => useStopwatch(1, onTimerEnd));

    act(() => {
      result.current.lapStopwatch();
    });

    act(() => {
      jest.advanceTimersByTime(1000);
    });

  });

});