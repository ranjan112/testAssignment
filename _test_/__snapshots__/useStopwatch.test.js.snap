// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTimer Jest Snapshot testing suite Matches DOM Snapshot 1`] = `<useStopwatch />`;
